playbooks/continuous_delivery/ci-cd-elements.adoc

```diff
@@ -0,0 +1,252 @@
+
+Environment independent validation steps:
+|===
+|*Validation step *|*Common tool/technology implementations*
+
+|Code static analysis|https://www.sonarqube.org/[SonarQube]
+|Code static security scanning|https://www.veracode.com/[Veracode]
+|Unit test|https://junit.org/junit5/[JUnit]
+|Image security scanning |https://www.redhat.com/en/technologies/management/cloudforms[Cloudforms] (with https://developers.redhat.com/blog/2016/05/02/introducing-atomic-scan-container-vulnerability-detection/[atomic scan]), https://www.twistlock.com/[Twistlock], https://www.blackducksoftware.com/[BlackDuck], https://www.aquasec.com/[Aquasec].
+|===
+
+Image security scanning is a new step introduced by container technology. It ensures the quality of images in terms of security vulnerabilities (https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures[CVEs]) +
+This process has two sides: on one side we need to scan the images to determine if there are known vulnerabilities in it, on the other side we need to prevent images that don’t pass the quality check from running as containers. This enforcement resides outside of the pipeline and can be configured in OpenShift via the https://docs.openshift.com/container-platform/3.7/admin_guide/image_policy.html[image admission control].  
+
+An example image scanning pipeline step can be found https://github.com/redhat-cop/openshift-image-signing-scanning[here].
+
+Environment dependent validation steps:
+
+|===
+|*Validation step*|*Common tool / technology implementations*
+
+|https://en.wikipedia.org/wiki/Integration_testing[Integration test]|Various tools and techniques. https://en.wikipedia.org/wiki/Behavior-driven_development[BDD] seems to be emerging for web and mobile apps.
+|https://martinfowler.com/bliki/CanaryRelease.html[Canary test]|Pipeline orchestrated and tested with integration testing tools.
+|https://en.wikipedia.org/wiki/Chaos_Monkey[Chaos test]|https://www.gremlin.com/[Gremlin], https://fabric8.io/guide/chaosMonkey.html[Fabric8 Chaos Monkey].
+|===
+
+Behavioral Driven Development (BDD) testing is a valuable option and should be used for testing web and mobile applications. +
+BDD test cases are created using a pseudo natural language (https://cucumber.io/docs/reference[Gerkin]), allowing them to be quickly and easily created by non-technical users. The test cases are translated into machine instructions at test runtime (a popular one is https://cucumber.io/[Cucumber]).  +
+The general architecture of BDD tests is illustrated in the following diagram:
+
+image::images/BDD.png[]
+
+The above architecture can be fully deployed and contained in OpenShift. Multiple combinations of browser/operating system and mobile device/operating system tests can be executed in parallel. The advent of Windows containers will provide additional opportunities for testing applications running on Internet Explorer and Edge.
```

Comment:
Where and how can it?

Reply:
we don't have guidance for this buy I know of a customer that is doing it.
I think we should create some playbook for it.